system/jlib/jptree.cpp

```diff
@@ -2981,6 +3014,156 @@ IPropertyTree *getXPathMatchTree(IPropertyTree &parent, const char *xpath)
     return matchTree;
 }
 
+void PTree::serializeAttributes(IBufferedSerialOutputStream &tgt) const
+{
+    Owned<IAttributeIterator> aIter = getAttributes();
+    if (aIter->first())
+    {
+        do
+        {
+            const char *name = aIter->queryName();
+            const char *value = aIter->queryValue();
+            size32_t nameLen = strlen(name);
+            size32_t valueLen = strlen(value);
+
+            tgt.put(sizeof(nameLen), &nameLen);
+            tgt.put(nameLen, name);
+            tgt.put(sizeof(valueLen), &valueLen);
+            tgt.put(valueLen, value);
+        }
+        while (aIter->next());
+    }
+    size32_t zero = 0;
+    tgt.put(sizeof(zero), &zero); // attribute terminator. i.e. blank attr name length.
+}
+
+void PTree::serializeSelf(IBufferedSerialOutputStream &tgt) const
+{
+    const char *_name = queryName();
+    const char *name = _name ? _name : "";
+    size32_t nameLen = strlen(name);
+
+    tgt.put(sizeof(nameLen), &nameLen);
+    tgt.put(nameLen, name);
+    tgt.put(sizeof(flags), &flags);
+    serializeAttributes(tgt);
+    if (value)
+        value->serializeToStream(tgt);
+    else
+    {
+        size32_t zero = 0;
+        tgt.put(sizeof(zero), &zero);
+    }
+}
+
+void PTree::serializeCutOff(IBufferedSerialOutputStream &tgt, int cutoff, int depth) const
+{
+    serializeSelf(tgt);
+
+    if (-1 == cutoff || depth<cutoff)
+    {
+        Owned<IPropertyTreeIterator> iter = getElements("*");
+        if (iter->first())
+        {
+            do
+            {
+                IPropertyTree *_child = &iter->query();
+                PTree *child = QUERYINTERFACE(_child, PTree); assertex(child);
+                child->serializeCutOff(tgt, cutoff, depth+1);
+            }
+            while (iter->next());
+        }
+    }
+    size32_t zero = 0;
+    tgt.put(sizeof(zero), &zero); // element terminator. i.e. blank child name length.
+}
+
+void PTree::serializeToStream(IBufferedSerialOutputStream &tgt) const
+{
+    serializeCutOff(tgt, -1, 0);
+}
+
+void PTree::deserializeFromStream(IBufferedSerialInputStream &src)
+{
+    deserializeSelfFromStream(src);
+
+    for (;;)
+    {
+        offset_t pos = src.tell();
+        size32_t nameLength;
```

Comment:
You should use MemoryAttr or similar to manage the allocation - not explicit malloc/free calls.